services/operations.go

```diff
@@ -305,11 +363,170 @@ func readOasisCurrencyNeg(amount *types.Amount) (*quantity.Quantity, error) {
+	amount, err := readOasisCurrencyNeg(ops[0].Amount)
+	if err != nil {
+		return nil, fmt.Errorf("invalid add escrow from amount: %w", err)
+	}
+
+	var escrowAccount staking.Address
+	if err = escrowAccount.UnmarshalText([]byte(ops[1].Account.Address)); err != nil {
+		return nil, fmt.Errorf("invalid add escrow to address (%s): %w", ops[1].Account.Address, err)
+	}
+	amount2, err := readOasisCurrency(ops[1].Amount)
+	if err != nil {
+		return nil, fmt.Errorf("invalid add escrow to amount: %w", err)
+	}
+	if amount.Cmp(amount2) != 0 {
+		return nil, fmt.Errorf("add escrow amounts differ between operations (from: %s to: %s)", amount, amount2)
+	}
+
+	escrow := staking.Escrow{
+		Account: escrowAccount,
+		Amount:  *amount,
+	}
+	return &escrow, nil
+}
+
+// getStakingReclaimEscrow decodes the Oasis staking reclaim escrow transaction
+// from the given Rosetta operations.
+func getStakingReclaimEscrow(ops []*types.Operation) (*staking.ReclaimEscrow, error) {
+	if ops[0].Amount != nil {
+		return nil, fmt.Errorf("invalid reclaim escrow's from amount (expected: nil): %s", ops[0].Amount.Value)
+	}
+
+	var escrowAccount staking.Address
+	if err := escrowAccount.UnmarshalText([]byte(ops[1].Account.Address)); err != nil {
+		return nil, fmt.Errorf("invalid reclaim escrow address (%s): %w", ops[1].Account.Address, err)
+	}
+	if ops[1].Amount != nil {
+		return nil, fmt.Errorf("invalid reclaim escrow's to amount (expected: nil): %s", ops[1].Amount.Value)
+	}
+	sharesRaw, ok := ops[1].Metadata[ReclaimEscrowSharesKey]
+	if !ok {
+		return nil, fmt.Errorf("reclaim escrow shares metadata not specified")
+	}
+	sharesStr, ok := sharesRaw.(string)
+	if !ok {
+		return nil, fmt.Errorf("malformed reclaim escrow shares metadata")
+	}
+	var shares quantity.Quantity
+	if err := shares.UnmarshalText([]byte(sharesStr)); err != nil {
+		return nil, fmt.Errorf("malformed reclaim escrow shares metadata (%s): %w", sharesStr, err)
+	}
+
+	reclaim := staking.ReclaimEscrow{
+		Account: escrowAccount,
+		Shares:  shares,
+	}
+	return &reclaim, nil
+}
+
+// checkSigner ensures the operation's signer address matches the given signer
+// address (if specified) and returns the operation's signer address.
+func checkOpSignerAddress(op *types.Operation, signerAddr string) (string, error) {
+	switch {
+	case signerAddr == "":
+		return op.Account.Address, nil
+	case signerAddr != op.Account.Address:
+		return "", fmt.Errorf("operation's address doesn't match signer's address (op: %s signer: %s)",
+			op.Account.Address,
+			signerAddr,
+		)
+	default:
+		return signerAddr, nil
+	}
+}
+
+const txnTypeUnknown = "unknown"
+
+// decodeOpsToTransaction decodes the Oasis transaction type from the given
+// Rosetta operations.
+func decodeOpsToTransaction(ops []*types.Operation) string {
+	switch {
+	case len(ops) == 1:
+		switch {
+		case ops[0].Type == OpBurn &&
+			ops[0].Account.SubAccount == nil:
+			return "staking_burn"
```

Comment:
All of these should be typed constants (e.g., return type should not be a generic string).

Reply:
Agreed. Done.